floodsub.go

```diff
@@ -324,6 +347,85 @@ func msgID(pmsg *pb.Message) string {
 	return string(pmsg.GetFrom()) + string(pmsg.GetSeqno())
 }
 
+// pushMsg pushes a message to a number of subscriptions, performing validation
+// as necessary
+func (p *PubSub) pushMsg(subs []*Subscription, src peer.ID, msg *Message) {
+	// we perform validation if _any_ of the subscriptions has a validator
+	// because the message is sent once for all topics
+	needval := false
+	for _, sub := range subs {
+		if sub.validate != nil {
+			needval = true
+			break
+		}
+	}
+
+	if needval {
+		// validation is asynchronous
+		// XXX vyzo: do we want a global validation throttle here?
+		go p.validate(subs, src, msg)
+		return
+	}
+
+	go func() {
```

Comment:
hrm... spawning a goroutine for each message being sent is a bit wonky. Maybe:
```go
select {
case p.sendMsg <- .....:
default:
  go func() {
    p.sendMsg <- ....
  }()
}
```

Reply:
Good point, we only have to spawn a new goroutine if it's a publish, in which case the push happens within the event loop.
I'll update.

Reply:
we are always in the event loop actually. I added a buffer so that we almost always avoid the goroutine.